📖 Add proposal for worker version pinning in Cluster API managed topologies (#13892)
kubernetes-sigs/cluster-api: docs/proposals/20260706-independent-worker-topology-versions.md

---

> title: Independent Worker Topology Versions

Reviewer: Wondering if this should be

```suggestion
title: Workers version pinning/manual version management in Cluster API managed topologies
```

---

<!--
 Short-form proposal draft following the Cluster API CAEP template.
 Target location once ready: docs/proposals/YYYYMMDD-independent-worker-topology-versions.md
 (rename YYYYMMDD to the date the proposal PR is first opened).
 Supersedes the earlier WIP proposal PR: https://github.com/kubernetes-sigs/cluster-api/pull/12698
-->

# Independent Worker Topology Versions

## Table of Contents

- [Glossary](#glossary)
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals/Future Work](#non-goalsfuture-work)
- [Proposal](#proposal)
  - [Overview](#overview)
  - [User Stories](#user-stories)
  - [API Changes](#api-changes)
  - [Behavior](#behavior)
  - [Security Model](#security-model)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Alternatives](#alternatives)
- [Upgrade Strategy](#upgrade-strategy)
- [Additional Details](#additional-details)
  - [Test Plan](#test-plan)
  - [Version Skew Strategy](#version-skew-strategy)
- [Implementation History](#implementation-history)

## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

> - **Worker Node Group Version**: An optional Kubernetes version set on an individual

Reviewer: nit: What about avoiding to introduce the idea of "Worker Node Group" and use a generic workers or if we want be more specific MachineDeployments/MachinePools (or MDs/MPs abbreviated)? (I think we should avoid introducing new terms for concept that already exists)

Note: after reading the entire proposal, I think we should drop Worker Node Group entirely.

Reviewer: Let's also please fix diagrams

  `MachineDeploymentTopology` or `MachinePoolTopology` in a Cluster's topology. When set, it
  overrides the cluster-wide topology version (`Cluster.spec.topology.version`) for that one
  worker node group, allowing that group to be versioned and upgraded independently.

## Summary

> Today, ClusterClass enforces a single Kubernetes version across the whole cluster topology.

Reviewer: ```suggestion
Today, Cluster API managed topologies enforces a single Kubernetes version across the whole cluster topology.
```
Note: This is the official name of the feature and also how we call it in the API (ClusterClass are only an implementation detail)

Please also replace ClusterClass with Cluster API managed topologies in the rest of the document

This proposal adds an optional per-worker-node-group version so that individual worker node
groups can be upgraded independently of the control plane and of each other, within the limits
of the [Kubernetes version skew policy](https://kubernetes.io/releases/version-skew-policy/).
Cluster admins keep the benefits of ClusterClass while gaining fine-grained control over
upgrade timing and blast radius.

## Motivation

ClusterClass ties every worker node group to a single cluster-wide topology version
(`Cluster.spec.topology.version`) and upgrades them under one coupled, controller-driven
sequence. While a worker node group is waiting to take its version step, the topology
controller defers that group's config changes and topology-driven scaling so they roll out
together with the version bump, avoiding a double rollout (the `IsPendingUpgrade` early-return
in [`reconcile_state.go`](https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/topology/cluster/reconcile_state.go);
machine remediation via MachineHealthCheck is deliberately exempt from this freeze). For
multi-minor upgrades the control plane and workers advance in lockstep — the control plane

> takes a step, then waits for the workers to catch up before taking the next

Reviewer: this doesn't seem entirely correct, because workers can skip minors

([`desired_state.go`](https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/topology/desiredstate/desired_state.go),
`ControlPlane.IsWaitingForWorkersUpgrade`). For large or heterogeneous clusters this coupling

> creates real friction:

Reviewer: Nit, What about

```suggestion
`ControlPlane.IsWaitingForWorkersUpgrade`). 

This approach works well for most cases, by removing most of the upgrade toil from users, but in  large or heterogeneous clusters this coupling creates some friction:
```

- A single worker group's rollout can take days, and no group can be upgraded independently of
  the cluster-wide version while it runs.
- Different workloads need different upgrade cadences (e.g. uninterruptible training jobs vs.
  customer-facing inference sharing one cluster via a custom scheduler).

> - The only workarounds today — dropping ClusterClass for raw MachineDeployments, or splitting

Reviewer: Let's add that we also have annotations allowing to defer/hold upgrade or to control upgrade sequence, but the downside of this approach is to stretch the upgrade duration and a lack of a proper granular version control

  into many clusters — cost guardrails, features, and money.

Decoupling worker-group upgrades from the global cluster version keeps the ClusterClass

> abstraction while removing this bottleneck.

Reviewer: ```suggestion
This proposal introduces the possibility to pin versions/manual version management for specific MD/MP, thus preserving all the benefits of Cluster API managed topologies abstraction while removing the friction described above.
```

### Goals

> - Preserve existing ClusterClass version behavior when the feature is not used.

Reviewer: ```suggestion
- Preserve existing Managed topologies version management when MD/MP version pinning is not used.
```

- Allow any single worker node group to be upgraded independently of the control plane and of

>   other worker node groups.

Reviewer: ```suggestion
- Allow pinning versions/manually manage version for each MD/MP.
```

- Keep normal operations (scaling, remediation, config changes) running on worker node groups

>   that are *not* the one being upgraded.

Reviewer: I'm not sure that I'm grasping this goal. Is this about pinned MD/MP or it is about MD/MP not yet upgrading?

- Support version skew between worker node groups and the control plane as allowed by the
  Kubernetes version skew policy, and keep the existing skew safety checks/guardrails.
- Keep version-aware patches predictable by using `builtin.machineDeployment.version` /
  `builtin.machinePool.version` as the source of truth for a group, even when it differs from
  `builtin.cluster.topology.version`.

### Non-Goals/Future Work

- **Allowing unsupported skew.** Only skew permitted by the Kubernetes version skew policy is
  allowed.
- **Version-aware built-in patches for independent groups.** Out of scope for this iteration;

>   candidate for future work.

Reviewer: I think this must be in scope, other wise patches for those nodes with stop working.
In other words, I think that patches should keep sending the specific version for a MD/MP, no matter if managed or pinned.

Author: I agree. This was carried over from the previous doc. We should definitely have this in scope.

- **Automatically dropping a worker-group version override.** Reverting a group back to
  cluster-managed versioning (e.g. once all groups have caught up to the control plane) is

>   deferred to future work to avoid adding another upgrade variant now.

Reviewer: Personally I think that pinning and unpinning must always be explicit actions, so I would drop line or frase it as a non goal

- **Changing rollout behavior for groups not using the feature.** Groups without an

>   independent version keep today's behavior; they are simply still included in skew checks.

Reviewer: ```suggestion
- **Changing rollout behavior for groups not using the feature.** Groups without an
  independent version keep today's behavior.
```
(the last part seems an implementation detail not relevant for stating the non-goal)

  CPV["Cluster.spec.topology.version = v1.32"] --> CPU["Control plane and cluster-managed worker groups upgrade to v1.32"]
  WV["MachineDeploymentTopology.version = v1.30, pinned group"] --> WU["Pinned group upgrades on its own schedule"]
  W -->|gates| CPU
  W -->|gates| WU
```

### User Stories

1. **Independent control-plane vs. worker upgrades.** As a cluster admin, I want to upgrade the
   control plane without forcing a (potentially disruptive) worker upgrade that must be
   coordinated with tenants.
2. **Per-group upgrade scheduling.** As a cluster admin, I have multiple worker groups for
   different purposes (e.g. inference vs. training) that need different upgrade timing in the
   same cluster.
3. **No cluster-wide freeze.** As a cluster admin with many/large worker groups, I don't want
   every other group frozen while one group's multi-day rollout runs.
4. **Smaller blast radius.** As a cluster admin, I want to upgrade one worker group first to
   validate the change before rolling it out cluster-wide.

### API Changes

Add an optional `version` field to `MachineDeploymentTopology` and `MachinePoolTopology`:

```go

> // version is the optional Kubernetes version for this worker node group.

Reviewer: ```suggestion
// version is the optional Kubernetes version for this MachineDeployment/MachinePool.
```
(please avoid introducing group in our API documentation

// When set, it overrides Cluster.spec.topology.version for this group only,
// enabling independent version management and upgrade scheduling.
//
// Skew between this version and the control plane version is enforced per the

> // Kubernetes version skew policy. If unset, the group uses Cluster.spec.topology.version.

Reviewer: ```suggestion
// Skew between this version and the control plane version is enforced according yo the
// Kubernetes version skew policy. 
// 
// If unset, the group uses Cluster.spec.topology.version.
// Once set, the field can be unset only if the MD/MP version is equal to Cluster.spec.topology.version.
```

//
// +optional
Version *string `json:"version,omitempty"`
```

The feature as a whole is guarded by a new feature gate — proposed name
`IndependentWorkerTopologyVersion` — that we will register (defaulted off) when we implement,
following the existing `ClusterTopology` gate in
[`feature/feature.go`](https://github.com/kubernetes-sigs/cluster-api/blob/main/feature/feature.go).
Cluster API gates behavior in controller and webhook code rather than via a field-level codegen
marker, so the gate is enforced by the topology controller and the validating webhooks (below);
the field is always present in the API and inert unless the gate is enabled. We will follow the
standard guide for adding a new field to an existing API version.

### Behavior

- **Feature gate off / created before enablement:** no independent versions; webhooks reject
  changes to worker-group versions and behavior is unchanged.

> - **Excluded from cluster-level rollouts:** a group with an independent version is skipped by

Reviewer: Is this 
```suggestion
- **Feature gate on / Excluded from cluster-level rollouts:** a group with an independent version is skipped by
```

  `Cluster.spec.topology.version` rollouts (and by chained upgrades), but is still evaluated by
  version-skew safety checks.

> - **Validating webhooks** enforce:

Reviewer: If we want to go in details like the following subpoints, may be we should split between new validation rules that applies to Cluster.spec.topology.version and validation rules for the new MD/MP.version field

  - allowable version skew between the control plane and every worker-group version (validated

>     against both current and incoming versions);

Reviewer: not sure I understand what incoming version means in this context, can you clarify?

  clear error via webhook validation and conditions, and let the admin align that group's
  version and remove the override to bring it back under cluster control.

## Alternatives

- **Raw MachineDeployments/MachinePools.** Attach non-topology MachineDeployments to a
  topology cluster. Loses ClusterClass guardrails (including skew safety) and worker classes.
- **One cluster per worker group.** Isolates upgrades but adds control-plane cost, complicates
  shared services and multi-tenant networking, requires cross-cluster config propagation, and
  still doesn't let you manage control-plane and worker versions independently within a
  cluster. Good fit for strong-isolation models, poor fit for shared single-cluster platforms.

## Upgrade Strategy

No impact on upgrading Cluster API itself. The behavior is opt-in: it requires the feature gate
to be enabled and the new API field to be set to have any effect.

## Additional Details

### Test Plan

- Existing behavior preserved when no worker-group versions are set (control plane + workers
  roll to the desired version).
- Setting a worker-group version for the first time triggers a rolling upgrade of only that

>   group; other groups and the control plane keep operating normally.

Reviewer: How can Setting a worker-group version for the first time can trigger an upgrade? (I assume we can only pin version when they are equal to the control plane version)

Author: Oh yeah. My understanding was incorrect. I was thinking that we could enable the pin AND upgrade at the same time. 

A pinned MD/MP version can never exceed the CP versions, and we can't downgrade. So we have to start with all versions being the same, which would not trigger an upgrade. So the steps are: 
1. Pin at the current version (enforced with webhooks) 
2. Upgrades happen later only when the pin is later raised to match the CP. 
3. unpin is explicit and only happens when pin is equal to `CLuster.spec.topology.version`.
